frontend/src/container/PanelWrapper/PiePanelWrapper.tsx

```diff
@@ -0,0 +1,220 @@
+
+import { PanelWrapperProps, TooltipData } from './panelWrapper.types';
+import { getLabel, lightenColor, tooltipStyles } from './utils';
+
+function PiePanelWrapper({
+	queryResponse,
+	widget,
+}: PanelWrapperProps): JSX.Element {
+	const [active, setActive] = useState<{
+		label: string;
+		value: string;
+		color: string;
+	} | null>(null);
+
+	const {
+		tooltipOpen,
+		tooltipLeft,
+		tooltipTop,
+		tooltipData,
+		hideTooltip,
+		showTooltip,
+	} = useTooltip<TooltipData>();
+
+	const { containerRef, TooltipInPortal } = useTooltipInPortal({
+		scroll: true,
+		detectBounds: true,
+	});
+
+	const panelData =
+		queryResponse.data?.payload?.data.newResult.data.result || [];
+
+	const isDarkMode = useIsDarkMode();
+
+	const pieChartData: {
+		label: string;
+		value: string;
+		color: string;
+	}[] = [].concat(
+		...(panelData
+			.map((d) =>
+				d.series?.map((s) => ({
+					label:
+						d.series?.length === 1
+							? getLabel(Object.values(s.labels)[0], widget.query, d.queryName)
+							: getLabel(Object.values(s.labels)[0], {} as Query, d.queryName, true),
+					value: s.values[0].value,
+					color: generateColor(
+						d.series?.length === 1
+							? getLabel(Object.values(s.labels)[0], widget.query, d.queryName)
+							: getLabel(Object.values(s.labels)[0], {} as Query, d.queryName, true),
+						themeColors.chartcolors,
+					),
+				})),
+			)
+			.filter((d) => d !== undefined) as never[]),
+	);
+
+	let size = 0;
+	let width = 0;
+	let height = 0;
+
+	const chartRef = useRef<HTMLDivElement>(null);
+	if (chartRef.current) {
+		const { offsetWidth, offsetHeight } = chartRef.current;
+		size = Math.min(offsetWidth, offsetHeight);
+		width = offsetWidth;
+		height = offsetHeight;
+	}
+	const half = size / 2;
+
+	const getFillColor = (color: string): string => {
+		if (active === null) {
+			return color;
+		}
+		const lightenedColor = lightenColor(color, 0.4); // Adjust the opacity value (0.7 in this case)
+		return active.color === color ? color : lightenedColor;
+	};
+
+	return (
+		<>
+			{!pieChartData.length && <div className="piechart-no-data">No data</div>}
+			{pieChartData.length > 0 && (
+				<>
+					<div className="piechart-container" ref={chartRef}>
+						<svg width={width} height={height} ref={containerRef}>
+							<Group top={height / 2} left={width / 2}>
+								<Pie
+									data={pieChartData}
+									pieValue={(data: {
+										label: string;
+										value: string;
+										color: string;
+									}): number => parseFloat(data.value)}
+									outerRadius={({ data }): number => {
+										if (!active) return half - 3;
+										return data.label === active.label ? half : half - 3;
+									}}
+									padAngle={0.02}
+									cornerRadius={3}
+									width={size}
+									height={size}
+								>
+									{
+										// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+										(pie) =>
+											pie.arcs.map((arc, index) => {
```

Comment:
If we have used this code from another source, can we link it here for future reference?

Reply:
Added